Fixes #25227 - facts:cleanup batch support

lzap opened the pull request:
Implementation of slower-paced facts cleanup which is more concurrent-friendly as we did for reports expiration.

https://community.theforeman.org/t/foreman-rake-facts-cleanup-batch-rolling-support/11425

lzap committed:
Fixes #25227 - facts:cleanup batch support

LangJV commented:
@lzap 

At first try this doesn't seem to "work" in that it's still processing as a single transaction.

Stack trace plus partial query here: https://pastebin.com/vDT49tjF I truncated the query portion because it ran off the screen/history again but there were far more than the 500 ID's I expected in that query.

Quick sanity proof that I indeed have the updated file(s) in question in place on the server I ran the rake job on:
```
[root@fmnapprh4 ~]# ls -al /usr/share/foreman/lib/tasks | grep facts.rake
-rw-r--r--. 1 root root  586 Oct 19 09:36 facts.rake
[root@fmnapprh4 ~]# cat /usr/share/foreman/lib/tasks/facts.rake | grep batch
  # foreman-rake facts:clean [batch_size=500]
    batch_size = ENV['batch_size'] ? ENV['batch_size'].to_i : 500
    count = FactCleaner.new.clean!(batch_size).deleted_count
[root@fmnapprh4 ~]# ls -al /usr/share/foreman/app/services | grep fact_cleaner.rb
-rw-r--r--.  1 root root  1898 Oct 19 09:39 fact_cleaner.rb
[root@fmnapprh4 ~]# cat /usr/share/foreman/app/services/fact_cleaner.rb | grep batch
  def clean!(batch_size = 500)
    delete_excluded_facts(batch_size)
    delete_orphaned_facts(batch_size)
  def delete_orphaned_facts(batch_size)
    FactName.leaves.includes(:fact_values).find_in_batches(:batch_size => batch_size) do |batch|
      fact_name_ids = batch.select { |fact| fact.fact_values.empty? }
    FactName.composes.find_in_batches(:batch_size => batch_size) do |batch|
      fact_name_ids = batch.select { |compose| compose.descendants.all? { |fact| leaf_ids.include?(fact.id) } }.map(&:id)
  def delete_excluded_facts(batch_size)
    FactName.unscoped.find_in_batches(:batch_size => batch_size) do |names_batch|
      fact_name_ids = names_batch.select! { |fact_name| fact_name.name.match(excluded_facts_regex) }.map(&:id)
[root@fmnapprh4 ~]#

```

LangJV commented:
This initial failure may have been 100% PEBCAK on my end- or i had my batch size too high with regard to the leaf's or "something" (10,000)... i lowered it to 100 and it is indeed processing now.... or at least - indicating it is:

2018-10-20T15:21:54 [D|app|] Removed 100 fact name records
2018-10-20T15:21:54 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records
2018-10-20T15:21:55 [D|app|] Removed 100 fact name records

I will follow up once I confirm it indeed "works" and "finishes"

LangJV commented:
This works in all 3 of my environments:

**Before:**
select count(*) from fact_names;
fact_names ‘7991166’

select count(*) from fact_values;
fact_values ‘8062049’

**After:**
select count(*) from fact_names;
fact_names ‘286204’

select count(*) from fact_values;
fact_values ‘8608332’

I vote this goes into 1.18.X and up - but ill leave it to the smarter minds to decide!

lzap commented:
Thank you this is helpful, so do you recommend also to lower batch size to 100 or is 500 good enough? Can you test with 500?

LangJV commented:
Hey @lzap - it was able to run successfully in my environment with batch_size several times since. Also indicating it was deleting in batches of 500 with no problem - query time to delete 500 records seemed to be consistently 2 seconds or (typically much) less.

With that in mind, I think 500 is a safe default :)

ShimShtein commented on app/services/fact_cleaner.rb:
```diff
@@ -1,80 +1,54 @@
 
-  def clean!
-    @deleted_count = delete_excluded_facts + delete_orphaned_facts
+  def clean!(batch_size = 500)
```
We can pass the batch size in the constructor and save it as instance variable.
It will make all method signatures cleaner: 
```suggestion
  def initialize(batch_size: 500)
    @batch_size = batch_size
  end
  
  def clean!
```

ShimShtein commented on lib/tasks/facts.rake:
```diff
@@ -6,12 +6,13 @@ This is useful when you used custom facts of different facts sourcers that you d
   task :clean => :environment do
-    puts 'Starting orphaned facts clean up...'
-    count = FactCleaner.new.clean!.deleted_count
+    batch_size = ENV['batch_size'] ? ENV['batch_size'].to_i : 500
```
I think it would be a bit cleaner to use official rake [parameters API](https://ruby.github.io/rake/doc/rakefile_rdoc.html#label-Tasks+that+Expect+Parameters):

``` ruby
task :clean[:batch_size] => :environment do
```

lzap replied:
It does not work on production installs for some reason.

ShimShtein replied:
From reading more threads about it, I see it's really sensitive to spaces (and for zsh, you have to escape `[]` too).

Let's leave it as `ENV`.

I have considered `OptionsParser`, but I think it would be over-complication for this use case.

ShimShtein reviewed (changes requested):
@lzap, Thanks for refactoring this.
Now it looks much better, especially with batch size feature.

A couple of small changes though.

lzap commented:
Rebased.

ShimShtein commented on app/services/fact_cleaner.rb:
```diff
@@ -1,80 +1,58 @@
   def clean!
-    @deleted_count = delete_excluded_facts + delete_orphaned_facts
+    @deleted_count = 0
+    delete_excluded_facts(@batch_size)
```
Do we need to pass `@batch_size` explicitly? We can just use it inside the inner methods.

lzap replied:
Oh no, a leftover.

ShimShtein commented:
Forgot to mention: All the output is sent to debug logger now, which is turned off by default in production. Can we output some progress to the user? Right now, it would look completely stuck until it's done.

lzap commented:
Rebased the leftovers. I changed the logging level on purpose, there is now much more logging (per batch), so verbosity would increase. I also don't think we need to give very detailed logging via INFO except the final "Deleted X facts" - the final number should be enough for starters I think.

lzap commented:
One more thing, I added `reorder` just before `find_in_batches` to get rid of Rails warning message about dropped ordering. Here it how a (debug) session looks like when there is nothing to delete:

```
Starting orphaned facts clean up
Cleaning facts mathing excluded pattern: (?-mix:(\A|.*::|.*_)lo(\Z|::.*)|(\A|.*::|.*_)en.*v.*(\Z|::.*)|(\A|.*::|.*_)usb.*(\Z|::.*)|(\A|.*::|.*_)vnet.*(\Z|::.*)|(\A|.*::|.*_)macvtap.*(\Z|::.*)|(\A|.*::|.*_)_vdsmdummy_(\Z|::.*)|(\A|.*::|.*_)veth.*(\Z|::.*)|(\A|.*::|.*_)docker.*(\Z|::.*)|(\A|.*::|.*_)tap.*(\Z|::.*)|(\A|.*::|.*_)qbr.*(\Z|::.*)|(\A|.*::|.*_)qvb.*(\Z|::.*)|(\A|.*::|.*_)qvo.*(\Z|::.*)|(\A|.*::|.*_)qr\-.*(\Z|::.*)|(\A|.*::|.*_)qg\-.*(\Z|::.*)|(\A|.*::|.*_)vlinuxbr.*(\Z|::.*)|(\A|.*::|.*_)vovsbr.*(\Z|::.*))
Cleaning leaf orphaned facts
Cleaning compose orphaned facts
Finished, cleaned 0 facts
```

mmoll commented:
@lzap please rebase :)

mmoll commented:
merged, díky @lzap!
